### ISSUE TITLE
Wavefunction state serialization capability needed

### ISSUE THREAD

**Issue by markdewing** [QMCPACK/qmcpack#3619]
A number of wavefunction parameters do not get saved during optimization (CI coeff, orbital rotations)

The current scheme updates XML nodes in `QMCCostFunctionBase::updateXmlNodes`.  That function needs to get updated for every new parameter.

See discussions in #444 and #2068

We want a general solution to saving variational parameters.

**Describe the solution you'd like**

The general solution is the ability to dump the wavefunction, including the parameters and other structure.
Desirable features:
* complete state and structure so the wavefunction can be re-used easily
  * As mentioned in issue 444, keeping the CI coefficients and Jastrow factors together is desirable.  The easy and common path should be to reuse all the variational parameters as a set.
* Both XML and HDF formats
  * XML is easy to read and modify for small cases
  * HDF is necessary for scaling to large numbers of parameters
* Implementation should be distributed to classes rather than centralized as it is now

**Describe alternatives you've considered**
***Generic object serialization***
In principle, object serialization is what we want.  Something like boost::serialization would do this.  However, the format is opaque.  I think we want to re-use the input file format for uniformity.

***Saving the list of variational parameters***
One easy approach would be to store just the variational parameters in a file.   This would require two files to reload the state: the input file with the wavefunction structural information plus the file with the updated variational parameters. For the common use case, this shouldn't be too confusing.  It would get a little trickier when modifying a wavefunction - the WF input file and VP file would need updated in sync.

This approach would be the fastest to implement.    The more I think about it, the more appealing this solution looks (at least as an interim solution).  

The code changes would be limited in scope. The list of variational parameters is centralized to the `OptVariables` list in the optimizer.  This would effectively be serializing this structure to/from disk.

***Saving the entire wavefunction***
The original design had the ideal of each class reading from XML (the `put` methods).  While appealing in theory, it has led to problems.  XML handling gets mixed into every class.  And the `put` method becomes the de-facto constructor which makes unit testing difficult.
The new approach is to separate the XML input into a separate class (the various `Input` classes). 

In order to make the input and output more agnostic towards XML and HDF, we need a more independent (declarative) description of the input data.  The `Input` classes seem like a natural place to add this data model, and to perform output.  For example, the `OhmmsAttbuteSet` and `ParameterSet` classes would be fairly straightforward to extend to writing XML, and to handling HDF.  

This approach requires defining the new model and implementing in all the needed classes before it can be used, so it will be the slowest to implement.

**Comment by prckent** [comment]
Thanks for making this Mark. Please can you arrange for some discussion if this issue does not get much traction. Happily this is not a performance critical part of the application so we can focus on what users need, will find practical and easiest.

I would add to your list a mention of all the orbital coefficients - in other words we need a scheme that can in principle handle all future possibilities. Perhaps this is what you meant.

I was although thinking of what you suggest as an interim solution. i.e. an HDF5 file containing _all_ the parameters that were passed in/out of the optimizer. There is already a framework to make a packed buffer of them, and this buffer is then all that needs to be saved or loaded. What the interface looks like will need some work. Can you propose some input file examples?

How to handle a second route of getting to the original wavefunction specifications, some mix of pure XML or mixed XML and HDF5, is also desirable. e.g. With the above route it might not be possible to change the optimized parameter sets. This is a problem for typical optimization strategies where we optimize a few sets of coefficients at once and gradually make the wavefunction more complex.

**Comment by markdewing** [comment]
The intent is for a solution that can handle all the variational parameters, and can be easily extended to handle new kinds of parameters.

**Comment by correaa** [comment]
I have been working in a HDF5 backend for Boost.Serialization. I don't remember the application I had in mind (probably reading WF coefficients for AFQMC) but due to lack of interest I suspended it. 

This would be useful if it is decided that a common serialization code can be used with XML archive Boost.Serialization (included in Boost), but I don't know if that would be breaking (backwards) change, since Boost.Serialization XML archives can be very particular.

**Comment by markdewing** [comment]
Three possible places to add a tag for a VP file:
1. at the top level (well, in `simulation` or `qmcsystem`)
2. in `wavefunction`
3. in the driver

The advantage of 2 is that it naturally associates the VP file with the wavefunction.

However, looking at the workflow in more detail, it might make more sense to keep it out of the wavefunction section.
The starting point for optimization is from parameters as given in wavefunction section, which is often a separate file.
The optimization loop runs, writing out the new parameters at each step.
Then VMC/DMC, possibly in a separate run.
Since the wavefunction tag/file contains the wavefunction structure, only the pointer to the vp tag/file needs to be added or updated. 
So 1 and 3 would fit this (only require main file to be updated)

The other point is the parameter values are used in the drivers.   The optimizer of course sets up the variables, and that would be a natural place to read in the file.  The other drivers use the variational parameters as loaded.  They would need added code to set up the single OptVariables list, read in the file, and set the parameters. (Likely this code could be extracted from the optimizer and become commonly used between the drivers.)

For 1 & 2 there would need to be added code to store and track the VP file until it is used in the drivers.

For these reasons, I lean towards putting a tag in the driver section.
It would look something like this
```
<qmc method="vmc">
   <parameter name="variational_parameter_file">fully_opt.s002.vp.h5</parameter>
</qmc>
```

**Comment by ye-luo** [comment]
Loading and dumping parameter to h5 is a needed feature of TWF.
The parameter sets of TWF in h5 needs to follow certain hierarchy. So I don't like serialization which wipes out TWF structure.
We may need to do additional specific operation after loading parameters so need virtual functions in TWF.

I think what you suggested is reasonable. Upon a request, drivers may load parameters from an external file.

**Comment by correaa** [comment]
Yes, serialization tends to flatten hierarchies (hence the name, serialization). Boost.Serialization can handle automatically a certain degree of hierarchy but it tends (by default) to be associated to the hierarchy implied by the class/member relationship. On the other hand this is not necessarily bad since it doesn't introduce more structure than the one in the code. If for any reason more hierarchy is required then, yes, serialization might not be the right approach.

**Comment by prckent** [comment]
I think the "driver being able to load and save parameters" route is the way to go. Hopefully the save functionality in the optimizer can be migrated up the class hierarchy. Some sanity checks will need to be added, e.g. no. of parameters in the input file match those currently active. The optimizer will be able to use this as the checkpoint format on each iteration as well. Overall this will also make the code much easier to restart.

**Comment by ye-luo** [comment]
If drivers detect "variational_parameter_file", they load and reset parameters.
The parameter output file name needs to following the scalar.dat filename conversion.
Regarding the name, I would like it indicating load but  "load_variational_parameter_file", is really too long.

**Comment by markdewing** [comment]
The optimizer already outputs an updated wavefunction using an ".opt.xml" suffix (that is the files have name "<project>.s*.opt.xml").  I'm thinking of using ".vp.xml" (or ".vp.h5") suffix for the files that store just the variational parameters.

I think "load_variational_parameter_file" is not too long (these tags will be read a lot more than written by hand).  One issue is the "load" at the beginning makes it look like a command.  If we want to indicate the values are being read rather then written, maybe "variational_parameter_file_to_load" ?

**Comment by ye-luo** [comment]
Adding "vp" avoids confusion with "opt". That is good. Since the output file carry "vp". We maybe can use "vp_file_to_load". Yes, I'd like to indicate it is reading rather than writing.

**Comment by prckent** [comment]
input_parameter_file ?

(Happy to get to this level of detail! In the manual we should note that we only support files from the same version of the code to give us freedom to change the format in the future.)
